library/core/src/fmt/mod.rs

```diff
@@ -612,19 +605,143 @@ impl<'a> Formatter<'a> {
+//         of the string.
+//
+//         Pieces can be 127 bytes at most. Longer pieces are split into multiple pieces (at utf-8
+//         boundaries).
+//
+//   - Placeholder:
+//         ┌──────────┬┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┬┄┄┄┄┄┄┄┄┄┄┄┬┄┄┄┄┄┄┄┄┄┄┄┬┄┄┄┄┄┄┄┄┄┄┄┐
+//         │0b10______│       flags       ┊   width   ┊ precision ┊ arg_index ┊ (e.g. b"\x82\x05\0")
+//         └────││││││┴┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┴┄┄┄┄┄┄┄┄┄┄┄┴┄┄┄┄┄┄┄┄┄┄┄┴┄┄┄┄┄┄┄┄┄┄┄┘
+//              ││││││        32 bit          16 bit      16 bit      16 bit
+//              │││││└─ flags present
+//              ││││└─ width present
+//              │││└─ precision present
+//              ││└─ arg_index present
+//              │└─ width indirect
+//              └─ precision indirect
+//
+//         Fully default placeholder, without any options:
+//         ┌──────────┐
+//         │0b10000000│ (b"\x80")
+//         └──────────┘
+//
+//         Placeholders (e.g. `{name}` in "hello {name}") are represented as a byte with the highest
+//         bit set, followed by zero or more fields depending on the flags set in the first byte.
+//
+//         The fields are stored as little endian.
+//
+//         The `flags` fields corresponds to the `flags` field of `FormattingOptions`.
+//         See doc comment of `FormattingOptions::flags` for details.
+//
+//         The `width` and `precision` fields correspond to their respective fields in
+//         `FormattingOptions`. However, if their "indirect" flag is set, the field contains the
+//         index in the `args` array where the dynamic width or precision is stored, rather than the
+//         value directly.
+//
+//         The `arg_index` field is the index into the `args` array for the argument to be
+//         formatted.
+//
+//         If omitted, the flags, width and precision of the default FormattingOptions::new() are
+//         used.
+//
+//         If the `arg_index` is omitted, the next argument in the `args` array is used (starting
+//         at 0).
+//
+//   - End:
+//         ┌───┐
+//         │ 0 │ ("\0")
+//         └───┘
+//         A single zero byte marks the end of the template.
+//
+//         (Note that a zero byte may also occur naturally as part of the string pieces or flags,
+//         width, precision and arg_index fields above. That is, the template byte sequence ends
+//         with a 0 byte, but isn't terminated by the first 0 byte.)
+//
 #[lang = "format_arguments"]
 #[stable(feature = "rust1", since = "1.0.0")]
 #[derive(Copy, Clone)]
 pub struct Arguments<'a> {
-    // Format string pieces to print.
-    pieces: &'a [&'static str],
+    template: NonNull<u8>,
+    args: NonNull<rt::Argument<'a>>,
+}
 
-    // Placeholder specs, or `None` if all specs are default (as in "{}{}").
-    fmt: Option<&'a [rt::Placeholder]>,
+/// Used by the format_args!() macro to create a fmt::Arguments object.
+#[doc(hidden)]
+#[rustc_diagnostic_item = "FmtArgumentsNew"]
```

Comment:
Is this a bootstrap thing?

Reply:
Why "New" then?

Reply:
"New" stands for `Arguments::new` I assume. The attribute used to be on an impl block with all `new*` functions.

Reply:
> Why "New" then?

Oh sorry, misunderstood your question. It was, and still is, called "FmtArgumentsNew" because these are the functions used to create a new fmt::Arguments. Those used to be `new_const`, `new_v1` and `new_v1_formatted`. Now those are `new`, `from_str` and `from_str_nonconst`.

These are special cased in const evaluation diagnostics to say "non-const formatting macro" rather than the (unstable/internal) function name.